backend/python-functions/data_ingest.py

```diff
@@ -1,12 +1,102 @@
+            # we need to keep track of them to stay below the Firestore batch limit
+            row_ops_count = 1
+            id = row['facility_id']
+
+            facility_doc_ref = facilities_collection.document(id)
+            facility_dict = {}
+
+            # are we updating or creating?
+            try:
+                fdoc_snapshot = facility_doc_ref.get()
+                assert fdoc_snapshot.exists
+                facility_dict.update(fdoc_snapshot.to_dict())
+            except AssertionError:
+                facility_dict.update({
+                    "createdAt": firestore.SERVER_TIMESTAMP,
+                })
+                row_ops_count += 1  # SERVER_TIMESTAMP is a separate operation
+
+            facility_dict.update({
+                "canonicalName": row['facility_name'],
+                "stateName": row['state'],
+                "facilityType": FACILITY_TYPE_MAPPING.get(row['facility_type'],
+                                                          row['facility_type']),
+            })
+
+            if row['capacity']:
+                facility_dict["capacity"] = int(row["capacity"])
+
+            if row['county'] and row['county'] != 'Not found':
+                facility_dict["county"] = re.sub(
+                    r' County$',  '', row['county'])
+
+            if row['population_year_updated'] and row['population']:
+                latest_population = {
+                    "date": datetime(int(row["population_year_updated"]), 1, 1, tzinfo=timezone.utc),
+                    "value": int(row["population"]),
+                }
+                # we have to be a bit careful not to obliterate existing data in the population array;
+                # generally this means looking for the item that corresponds to our input
+                # and leaving any other items untouched
+                if 'population' in facility_dict:
+                    # try to find an existing record matching this date
+                    # (we only care about the year since that's all we have in the data source)
+                    matching_record = next(
+                        (r for r in facility_dict['population']
+                            if r['date'].year == latest_population['date'].year),
+                        None)
+                    # append or update accordingly
+                    if matching_record is None:
+                        facility_dict['population'].append(latest_population)
+                    else:
+                        matching_record['value'] = latest_population['value']
+                else:
+                    facility_dict['population'] = [latest_population]
+
+            batch_counter += row_ops_count
+            # if we have hit our limit, start a new batch before proceeding
+            if batch_counter >= BATCH_SIZE:
+                batch.commit()
+                batch = fs_client.batch()
+                batch_counter = row_ops_count
```

Comment:
I don't immediately understand why we are resetting batch_counter to the ever-increasing row_ops_count at this point, but my brain may just be friend. Walk me through what's happening here?

Reply:
`row_ops_count` is not ever-increasing, it gets reset to `1` when we start a new row (on L33).

Because we are not necessarily incrementing by 1 every time, it's possible to overshoot the limit. So the logic here is, if processing this row _would_ put us at or over the limit, we start a new batch first and then handle this row as part of that new batch. Thus the initial number of ops in our new `batch_counter` is the number we will need to process the current row.

It's not the most intuitive thing, I agree ... maybe it would be a little clearer if we only did this once we had exceeded the batch size? And/or I could just write a longer comment about all this nonsense

Reply:
No it makes more sense now that I look at it again. I think it's fine as-is!